### PR TITLE
feat: validating asset scrap date

### PR DESCRIPTION
**Pull request by rahulgupta8848**
## Background/Context

1. When we select Asset Scrapping, the system by default takes the current date and scraps the asset as on the date
2. In case the company wants to scrap their asset with backdated effect, then that cannot be achieved in the present system as there is no option to select the scrap date anywhere.
3. Also, there is no restriction on scrapping the asset even if depreciation entries are posted

## Solution Summary

1. If a user wants to scrap an asset, then the system will ask for a scrapping date (as shown in the below screenshot).
2. According to the date entered by the user, the asset will be scrapped on that date with back dated effect.
4. The scrapping date cannot be before the purchase date or it cannot be after the current date. Also, If the depreciation entries are passed, asset can only be scrap after the last depreciation date and not before it. In such cases depreciation entry needs to be manually cancelled first and then scrapping can be possible.

image

#no-docs

### CODE REVIEW

File: erpnext/assets/doctype/asset/asset.py [hunk 1]
```diff
@@ -1269,3 +1269,48 @@ def add_reference_in_jv_on_split(entry_name, new_asset_name, old_asset_name, dep
...
+		last_depreciated = None
+		for entry in depreciation_entries:
+			if entry["journal_entry"]:
+				last_depreciated = entry
+			else:
+				if (
+					last_depreciated
+					and scrap_date < last_depreciated["schedule_date"]
+					and scrap_date > purchase_date
+				):
```
Comment (nabinhait) [inline erpnext/assets/doctype/asset/asset.py]: You can get last depreciation date simply by using `get_value`:

```
last_booked_depreciation_date = frappe.db.get_value(
		"Depreciation Schedule",
		{
			"parent": asset,
			"docstatus": 1,
			"journal_entry": ["!=", ""],
		},
		"schedule_date",
		order_by="schedule_date desc",
	)
```

File: erpnext/assets/doctype/asset/depreciation.py [hunk 2]
```diff
@@ -431,15 +431,15 @@ def get_comma_separated_links(names, doctype):
...
+def scrap_asset(asset_name, scrap_date=None):
 	asset = frappe.get_doc("Asset", asset_name)
 
 	if asset.docstatus != 1:
 		frappe.throw(_("Asset {0} must be submitted").format(asset.name))
 	elif asset.status in ("Cancelled", "Sold", "Scrapped", "Capitalized", "Decapitalized"):
 		frappe.throw(_("Asset {0} cannot be scrapped, as it is already {1}").format(asset.name, asset.status))
 
-	date = today()
+	date = scrap_date if scrap_date else today()
```
Comment (nabinhait) [inline erpnext/assets/doctype/asset/depreciation.py]:
```suggestion
	date = scrap_date or today()
```

File: erpnext/assets/doctype/asset/asset.js [hunk 3]
```diff
@@ -799,15 +799,35 @@ frappe.ui.form.on("Asset Finance Book", {
 });
 
 erpnext.asset.scrap_asset = function (frm) {
-	frappe.confirm(__("Do you really want to scrap this asset?"), function () {
-		frappe.call({
-			args: {
-				asset_name: frm.doc.name,
+	var scrap_dialog = new frappe.ui.Dialog({
+		title: "Enter date to scrap asset",
```
Comment (nabinhait) [inline erpnext/assets/doctype/asset/asset.js]:
```suggestion
		title: __("Enter date to scrap asset"),
```

File: erpnext/assets/doctype/asset/asset.js [hunk 4]
```diff
@@ -799,15 +799,35 @@ frappe.ui.form.on("Asset Finance Book", {
...
+		primary_action_label: "Submit",
+		primary_action(values) {
+			frappe.call({
+				args: {
+					asset_name: frm.doc.name,
+					scrap_date: values.scrap_date,
+					purchase_date: frm.doc.purchase_date,
+					calculate_depreciation: frm.doc.calculate_depreciation,
+				},
+				method: "erpnext.assets.doctype.asset.asset.validate_asset_scrap_date",
```
Comment (nabinhait) [inline erpnext/assets/doctype/asset/asset.js]: We should call the `scrap_asset` method from here as it is the primary task and `validate_asset_scrap_date` should be called from inside the `scrap_asset function`

File: erpnext/assets/doctype/asset/test_asset.py [hunk 5]
```diff
@@ -221,6 +221,31 @@ def test_scrap_asset(self):
...
+			if entry["journal_entry"]:
+				last_depreciated_row = entry
+
+		if asset.disposal_date > asset.purchase_date:
+			if last_depreciated_row:
+				disposal_date = frappe.utils.getdate(asset.disposal_date)
+				last_schedule_date = frappe.utils.getdate(last_depreciated_row["schedule_date"])
+				today_date = frappe.utils.getdate(nowdate())
+
+				self.assertTrue(
```
Comment (nabinhait) [inline erpnext/assets/doctype/asset/test_asset.py]: This does not make any sense, you should check for `assertRaise` by defining an error class to it.

File: erpnext/assets/doctype/asset/test_asset.py [hunk 6]
```diff
@@ -221,6 +221,31 @@ def test_scrap_asset(self):
...
+			"Depreciation Schedule",
+			filters={"parent": asset_depriciation},
+			fields=["schedule_date", "journal_entry"],
+			order_by="schedule_date",
+		)
+
+		last_depreciated_row = None
+		for entry in depreciation_entries:
+			if entry["journal_entry"]:
+				last_depreciated_row = entry
```
Comment (nabinhait) [inline erpnext/assets/doctype/asset/test_asset.py]: Like before, you can get last depreciation date using `get_value` here as well.

**Review by nabinhait — CHANGES_REQUESTED** [review]

**Comment by rahulgupta8848** [comment]
@nabinhait i have implemented all the suggested changes please look into it

**Review by nabinhait — APPROVED** [review]

**Comment by khushi8112** [comment]
Closes https://github.com/frappe/erpnext/issues/42514

**Pull request merged**